dali/pipeline/operator/arg_helper.h

```diff
@@ -15,94 +15,170 @@
+  return *reinterpret_cast<const mat<N, M>*>(view.data);
+}
+
+template <int N, int M>
+mat<N, M> as_mat(TensorView<StorageCPU, const float, DynamicDimensions> view) {
+  return as_mat<N, M>(view.to_static<2>());
+}
+
+template <int ndim>
+struct ArgShapeFromSize {
+  TensorShape<ndim> operator()(int64_t size) const {
+    throw std::logic_error(make_string("Cannot infer an ", ndim, "D shape from a flat size."));
+    return {};
   }
+};
 
-  ArgValue &operator=(ArgValue &&) = default;
-  inline ArgValue &operator=(const ArgValue &other) {
-    name_ = other.name_;
-    is_set_ = other.is_set_;
-    gpu_dirty_ = true;
-    gpu_.reset();
-    value_  = other.value_;
-    arg_input_ = other.arg_input_;
-    return *this;
+template <>
+struct ArgShapeFromSize<0> {
+  TensorShape<0> operator()(int64_t size) const {
+    DALI_ENFORCE(size == 1, make_string("Expected a scalar argument, got ", size, " values"));
+    return {};
   }
+};
+
+template <>
+struct ArgShapeFromSize<1> {
+  TensorShape<1> operator()(int64_t size) const {
+    return { size };
+  }
+};
 
-  inline bool IsInput() const noexcept { return arg_input_ != nullptr; }
-  inline bool IsSet() const noexcept { return is_set_; }
 
-  template <int ndim = DynamicDimensions>
-  inline const TensorView<StorageCPU, T, ndim> Tensor(int sample_index) const {
-    ASSERT(IsInput());
-    return view<T, ndim>(arg_input_[sample_index]);
+template <typename T, int ndim = 0>
+class ArgValue {
+ public:
+  using TLV = TensorListView<StorageCPU, const T, ndim>;
+  using TV = TensorView<StorageCPU, const T, ndim>;
+
+  ArgValue(std::string arg_name, const OpSpec &spec)
+      : arg_name_(std::move(arg_name)) {
+    has_arg_const_ = spec.HasArgument(arg_name_);
+    has_arg_input_ = spec.HasTensorArgument(arg_name_);
+    assert(!(has_arg_const_ && has_arg_input_));
   }
 
-  inline const T &Value() const {
-    return value_;
+  bool IsDefined() const {
+    return has_arg_const_ || has_arg_input_;
   }
 
-  inline const string &Name() const noexcept {
-    return name_;
+  bool IsConstant() const {
+    return has_arg_const_;
   }
 
-  inline const T &operator[](int sample_index) const {
-    if (IsInput()) {
-#if DALI_DEBUG
-      DALI_ENFORCE(sample_index < static_cast<int>(arg_input_->ntensor()));
-#endif
-      return (*arg_input_)[sample_index].data<T>()[0];
+  bool IsArgInput() const {
+    return has_arg_input_;
+  }
+
+  void Acquire(const OpSpec &spec, const ArgumentWorkspace &ws, int nsamples,
+               const TensorShape<ndim> &expected_shape) {
+    if (has_arg_input_) {
+      view_ = view<const T, ndim>(ws.ArgumentInput(arg_name_));
+      DALI_ENFORCE(is_uniform(view_.shape) && expected_shape == view_.shape[0],
+        make_string("Expected uniform shape for argument \"", arg_name_,
+                    "\" but got shape ", view_.shape));
     } else {
-      return  value_;
+      if (ndim == 0) {
+        data_ = {spec.GetArgument<T>(arg_name_)};
```

Comment:
What happens here if someone passes us a vector for this argument? Will it throw an error as the underlying type would be a vector<T> instead of T? I don't see this scenario in tests.

Reply:
It'll fail. I'll add that to the tests